cty/convert/conversion.go

```diff
@@ -43,14 +43,14 @@ func getConversion(in cty.Type, out cty.Type, unsafe bool) conversion {
 			out = out.WithoutOptionalAttributesDeep()
 
 			if !isKnown {
-				return cty.UnknownVal(out), nil
+				return cty.UnknownVal(dynamicReplace(in.Type(), out)), nil
```

Comment:
@apparentlymart - I specifically wanted to fix Null values with this, but I think it makes sense that unknown and null values behave the same. In case, we don't want that let me know and I'll revert the change to this line.